### PR TITLE
Enable bzlmod

### PR DESCRIPTION
**Pull request by dstufft**
This makes it possible to consume ``hermetic_cc_toolchain`` via ``MODULE.bazel``, in a way that is similar to how it is consumed with ``WORKSPACE``.

This doesn't setup the infrastructure to publish to BCR, so it won't make it available by default. 

This can get used in a ``MODULES.bazel`` like so:

```
bazel_dep(name = "hermetic_cc_toolchain", version = "2.0.0")

toolchains = use_extension("@hermetic_cc_toolchain//toolchain:ext.bzl", "toolchains")
use_repo(toolchains, "zig_sdk")

register_toolchains(
    "@zig_sdk//toolchain:linux_amd64_gnu.2.28",
    "@zig_sdk//toolchain:linux_arm64_gnu.2.28",
    "@zig_sdk//toolchain:darwin_amd64",
    "@zig_sdk//toolchain:darwin_arm64",
    "@zig_sdk//toolchain:windows_amd64",
    "@zig_sdk//toolchain:windows_arm64",
)

archive_override(
    module_name = "hermetic_cc_toolchain",
    urls = [
        "https://github.com/uber/hermetic_cc_toolchain/archive/c602db477e16f1cb10bfbd6c68013604500c4fbe.tar.gz",
    ],
    strip_prefix = "hermetic_cc_toolchain-c602db477e16f1cb10bfbd6c68013604500c4fbe",
    integrity = "sha256-g0WeBysSEkiHEdWmGkUnvTJEGwvqa/N+FDZDNPhUcoM=",
)
```

If it gets published to BCR, the ``archive_override`` would be able to be omitted.

Note: This moves the gazelle definitions into ``//tools/gazelle``, because bzlmod needs to be able to read the ``BUILD`` at the root of the repository, and the dependency on gazelle there would create a hard dependency on gazelle. Moving them down fixes that, but ``bazel //:gazelle`` still works thanks to the aliases.

### CODE REVIEW

**Comment by motiejus** [comment]
Looks good to me, will defer to @linzhp for a proper review (I did not yet take the time to make myself familiar to bzlmod).

**Comment by motiejus** [comment]
Question about bzlmod: currently our released builds do not have their version number in the archive. That way we are able to create the archive (and know it's checksum) before cutting the release tarball. It makes it quite convenient -- we have a [single commit](https://github.com/uber/hermetic_cc_toolchain/commit/9f2456cb09fc4be74621e6b84aecae44ce0530d5) that both updates the version number and the hash of the release tarball.

Does the version number *have* to be in `MODULE.bazel`? If yes, I will need to change the release process again, so it has two commits: one for bumping the version number, the other for changing the hashes.

**Review by linzhp — APPROVED** [review]

**Comment by dstufft** [comment]
I think you could avoid committing it if you template the version number so that the correct version number is in the released artifacts.

Also note, you can release this to the BCR automatically using [Publish to BCR](https://github.com/bazel-contrib/publish-to-bcr) workflow. I'm pretty sure that this uses the GitHub releases, so whenever a new release is published, a PR is automatically opened up to the BCR to incorporate the updated module into registry.bazel.build.

I can't set that up since it relies on cloning a repo and installing an app in those two repos, but it would also be great to get that setup at some point as well.

**Comment by motiejus** [comment]
> I think you could avoid committing it if you template the version number so that the correct version number is in the released artifacts.

Thanks. I will change the releaser to make this work later today.

> Also note, you can release this to the BCR automatically using [Publish to BCR](https://github.com/bazel-contrib/publish-to-bcr) workflow. I'm pretty sure that this uses the GitHub releases, so whenever a new release is published, a PR is automatically opened up to the BCR to incorporate the updated module into registry.bazel.build.

I'll poke at this after I fix the release tool.

> I can't set that up since it relies on cloning a repo and installing an app in those two repos, but it would also be great to get that setup at some point as well.

**Review by motiejus — APPROVED** [review]
#97 is non-blocking this PR, so we're fine.

Can you rebase? That way CI will run on a working main.

**Comment by motiejus** [comment]
Not sure why CI does not trigger. Hopefully rebase+re-approval will help.

**Review by motiejus — APPROVED** [review]

**Comment by motiejus** [comment]
Unfortunately, I am unable to push this through, as my knowledge on buildkite and that area is lacking. Will leave to @linzhp to handle it.

**Pull request merged**
